barretenberg/cpp/src/barretenberg/plonk_honk_shared/composer/permutation_lib.hpp

```diff
@@ -45,40 +45,77 @@ struct cycle_node {
  *
  */
 struct permutation_subgroup_element {
-    uint32_t row_index = 0;
-    uint8_t column_index = 0;
+    uint32_t row_idx = 0;
+    uint8_t column_idx = 0;
     bool is_public_input = false;
     bool is_tag = false;
 };
 
+/**
+ * @brief Stores permutation mapping data for a single wire column
+ *
+ */
+struct Mapping {
+    std::shared_ptr<uint32_t[]> row_idx; // row idx of next entry in copy cycle
+    std::shared_ptr<uint8_t[]> col_idx;  // column idx of next entry in copy cycle
+    std::shared_ptr<bool[]> is_public_input;
+    std::shared_ptr<bool[]> is_tag;
+    size_t _size = 0;
+
+    Mapping() = default;
+
+    size_t size() const { return _size; }
+
+    Mapping(size_t n)
+        : row_idx(_allocate_aligned_memory<uint32_t>(n))
+        , col_idx(_allocate_aligned_memory<uint8_t>(n))
+        , is_public_input(_allocate_aligned_memory<bool>(n))
+        , is_tag(_allocate_aligned_memory<bool>(n))
+        , _size(n)
+    {}
+};
+
 template <size_t NUM_WIRES, bool generalized> struct PermutationMapping {
-    using Mapping = std::array<std::vector<permutation_subgroup_element>, NUM_WIRES>;
-    Mapping sigmas;
-    Mapping ids;
+    std::array<Mapping, NUM_WIRES> sigmas;
+    std::array<Mapping, NUM_WIRES> ids;
 
     /**
      * @brief Construct a permutation mapping default initialized so every element is in a cycle by itself
      *
      */
     PermutationMapping(size_t circuit_size)
     {
-
         PROFILE_THIS_NAME("PermutationMapping constructor");
 
-        for (uint8_t col_idx = 0; col_idx < NUM_WIRES; ++col_idx) {
-            sigmas[col_idx].reserve(circuit_size);
```

Comment:
I thought reserve doesn't actually initialize memory, just allocates it

Reply:
yeah that's true, I thought I worked out that there was an extra zero init happening somewhere but now I dont see it. Maybe it was just that with this old approach is was very difficult to implement multithreading without one of either zero initializing or copy overhead